### PR TITLE
[PyTorch] Avoid double indirection in MaybeOwned's borrowed state

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55940 [PyTorch] Always update op.current_dtype in TensorIteratorBase::set_output
* #55420 [PyTorch] Remove non-const TensorIterator::tensor() method
* #55258 [PyTorch] Format generated structured kernels code better
* #55351 [PyTorch] Fix const correctness for resize native functions
* **#55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state**
* #55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned

This diff introduces a traits class that tells `MaybeOwned` how
to borrow a specific type. While it is still capable of handling a
generic `T` by storing `const T*` and how to do so is shown in a
comment, it is not committed in live code because it is not needed.

Instead, we have specific traits implementations for
`c10::intrusive_ptr<T>` and `Tensor` that implement the borrowed state
as just a plain old `c10::intrusive_ptr<T>` or `Tensor` (respectively)
that we manipulate to avoid reference counting operations. We do this
entirely with public API to `c10::intrusive_ptr<T>` and could do
likewise with `Tensor`, but (as comments in the code explain) adding a
private constructor to `MaybeOwnedTraits<Tensor>` allowed additional
performance optimization.

This representation of `MaybeOwned` seems to be more efficient than
the generic `T-or-pointer-to-const-T` representation. Intuitively, we
avoid a double indirection at minimal cost vs the previous
implementation. It *also* seems to be more efficient than the pointer
tagging representation I sent out as #55555; apparently, having the
extra word for a flag is much cheaper than the masking operands for
pointer tagging and the same double indirection as the generic
representation.

In particular, this seems to have the same *effect* as the
`TensorHandle` idea we've discussed internally (a hypothetical class
like `Tensor` that wraps a raw `TensorImpl*` and shares the generated
methods of `Tensor` so that everything still works), but you have to
be explicit about borrowing and use pointer syntax to get the
effect. Unlike `TensorHandle`, you can use it as internal state in a
class and "upgrade" from a borrow to an owned `Tensor` derived from
your original borrow if necessary.

Note that this is just a representational change and it still has the
same semantics: you need to keep the T you borrowed from around!

Differential Revision: [D27679723](https://our.internmc.facebook.com/intern/diff/D27679723/)

### CODE REVIEW

File: aten/src/ATen/templates/TensorBody.h [hunk 1]
```diff
@@ -86,6 +86,16 @@ inline bool variable_excluded_from_dispatch() {
 // Note that Tensor can also be NULL, i.e. it is not associated with any underlying TensorImpl, and
 // special care must be taken to handle this.
 class TORCH_API Tensor {
+ private:
+  struct unsafe_borrow_t { explicit unsafe_borrow_t() = default; };
```
Comment (ezyang) [inline aten/src/ATen/templates/TensorBody.h:90]: heh, you like this style of named constructor better than the enums we've been using? XD

**Review by ezyang — APPROVED** [review]
I'm pretty pumped about this. The pointer syntax downside seems like a good tradeoff here, because MaybeOwned is supposed to be for specialized performance sensitive use and not general consumption.

**Comment by ezyang** [comment]
> Note that this is just a representational change and it still has the
> same semantics: you need to keep the T you borrowed from around!

@swolchok This isn't "strictly" true anymore for Tensor, right? Because so long as you know that the TensorImpl in question is live somewhere, you don't actually need the literal Tensor you borrowed from live (previously you did need to keep it live!)

**Review by bhosmer — APPROVED** [review]

**Comment by swolchok** [comment]
> > Note that this is just a representational change and it still has the
> > same semantics: you need to keep the T you borrowed from around!
> 
> @swolchok This isn't "strictly" true anymore for Tensor, right? Because so long as you know that the TensorImpl in question is live somewhere, you don't actually need the literal Tensor you borrowed from live (previously you did need to keep it live!)

sure, it *could* support a slightly more complicated semantics like that, but "the original Tensor needs to be alive" is simple and correct, whereas unless your code is silly, you are unlikely to lose the original Tensor and still be sure that the TensorImpl really alive. (In other words, if you know that the Tensor is +2, why didn't you change the code to keep it only +1?)

**Pull request closed without merging**